pkg/crc/pullsecret/pullsecret.go

```diff
@@ -0,0 +1,169 @@
+kind: Secret
+metadata:
+  name: pull-secret
+  namespace: openshift-config
+type: kubernetes.io/dockerconfigjson`
+
+	// This command timeout in 80 secs if not able to replace the pull secret.
+	replacePullSecretCmd = `timeout 80 bash -c 'until oc --config /tmp/kubeconfig replace -f /tmp/pull-secret.yaml 2>/dev/null 1>&2; \
+do echo "Waiting for recovery apiserver to come up."; sleep 1; done'`
+
+	// This command is used to update the clusterID using random uuid.
+	updateClusterIdCmd = `oc --config /tmp/kubeconfig patch clusterversion version -p '{"spec":{"clusterID":"%s"}}' --type merge`
+
+	// This command make sure we stop the kubelet and clean up the pods
+	// We also providing a 2 sec sleep so that stop pods get settled and
+	// ready for remove. Without this 2 sec time sometime it happen some of
+	// the pod are not completely stopped and when remove happen it will throw
+	// an error like below.
+	// remove /var/run/containers/storage/overlay-containers/97e5858e610afc9f71d145b1a7bd5ad930e537ccae79969ae256636f7fb7e77c/userdata/shm: device or resource busy
+	stopAndRemovePodsCmd = `bash -c 'sudo crictl stopp $(sudo crictl pods -q) && sleep 2 &&\
+sudo crictl rmp $(sudo crictl pods -q)'`
+)
+
+func AddPullSecretAndClusterID(driver drivers.Driver, pullSec string, kubeconfigFilePath string) error {
+	// Add the kubeconfig File to the Instance.
+	if err := addKubeconfigFileToInstance(driver, kubeconfigFilePath); err != nil {
+		return err
+	}
+
+	// Add the Pull secret kubernetes resource definition to Instance
+	if err := addpullSecretSpecToInstance(driver, pullSec); err != nil {
+		return err
+	}
+
+	// Replace user pull secret and add cluster ID
+	if err := setPullSecretAndClusterID(driver); err != nil {
+		return err
+	}
+
+	// Add user pull secret to the instance
+	if err := addpullSecretToIntanceDisk(driver, pullSec); err != nil {
+		return err
+	}
+	return nil
+}
+
+func addpullSecretSpecToInstance(driver drivers.Driver, pullSec string) error {
+	base64OfPullSec := base64.StdEncoding.EncodeToString([]byte(pullSec))
+	output, err := drivers.RunSSHCommandFromDriver(driver, fmt.Sprintf("cat <<EOF | tee /tmp/pull-secret.yaml\n%s\nEOF", fmt.Sprintf(pullSecret, base64OfPullSec)))
+	if err != nil {
+		return err
+	}
+	logging.DebugF("Output is : %s", output)
+	return nil
+}
+
+func addpullSecretToIntanceDisk(driver drivers.Driver, pullSec string) error {
+	output, err := drivers.RunSSHCommandFromDriver(driver, fmt.Sprintf("cat <<EOF | sudo tee /var/lib/kubelet/config.json\n%s\nEOF", pullSec))
+	if err != nil {
+		return err
+	}
+	logging.DebugF("Output is : %s", output)
+	return nil
+}
+
+func addKubeconfigFileToInstance(driver drivers.Driver, kubeconfigFilePath string) error {
+	kubeconfig, err := ioutil.ReadFile(kubeconfigFilePath)
+	if err != nil {
+		return err
+	}
+	_, err = drivers.RunSSHCommandFromDriver(driver, fmt.Sprintf("cat <<EOF | tee /tmp/kubeconfig\n%s\nEOF", string(kubeconfig)))
+	if err != nil {
+		return err
+	}
+	return nil
```

Comment:
I'm wondering if we should introduce a copyFileToInstance()/copyDataToInstance helper as we duplicate this a few times.